### DISCUSSION TITLE
Optimization including stability derivatives

### DISCUSSION THREAD

**Discussion by Daniel-E-B** [peterdsharpe/AeroSandbox#78] [Q&A] [answered]
I am trying to optimize a flying wing for fabrication on a CNC hotwire. I can therefore only vary:
two chords,
span,
twist,
dihedral,
sweep.
Area is fixed.

I would like to ensure the static stability of the model but found this not to work. I am curious what I am doing wrong.
I can constrain `aero["Cm"] = -0.02` and nothing else. If I add any constraints (even things like <10 or ==0) for `Cl` or `Cn` or change my `Cm` constraint I get an `Infeasible_Problem_Detected` error.

Objective function is currently max L/D. My code is based on the 3D VLM example with all surfaces except the main wing removed.

Any ideas?

**Comment by Daniel-E-B** [comment, marked as answer]
I think I confused Cm with Cma, etc. Ignore this question

**Reply by sumitkumarsahoo2017** [reply]
Can you tell here how to calculate Cma and use it as a constraint?

**Reply by peterdsharpe** [reply]
Here's a tutorial!

https://github.com/peterdsharpe/AeroSandbox/blob/master/tutorial/10%20-%20Miscellaneous/02%20-%20Constraining%20Gradients.ipynb

Could also finite-difference Cma and constrain that.

**Comment by peterdsharpe** [comment]
Hi Daniel!

Glad to hear you found a fix. 

As a side note: if you are optimizing for max L/D, I might suggest using the `asb.AeroBuildup` aerodynamics engine, rather than `asb.VortexLatticeMethod` - AeroBuildup will include profile drag, which is important here. API is identical between the two, so you can probably just do a drop-in change.

EDIT: section below is irrelevant following AeroSandbox v4.1.0+

> If you're using `AeroBuildup`, you can further improve accuracy by [downloading XFoil](https://github.com/peterdsharpe/AeroSandbox/blob/master/tutorial/06%20-%20Aerodynamics/03%20-%20Interfaces%20to%20External%20Aerodynamics%20Tools/02%20-%20XFoil.ipynb) and auto-generating airfoil polars (see [Line 12 and 13 here](https://github.com/peterdsharpe/AeroSandbox/blob/master/tutorial/03%20-%20Trajectory%20Optimization%20and%20Optimal%20Control/03%20-%20The%20AeroSandbox%20Dynamics%20Stack/cessna152.py#L12); `Airfoil.generate_polars(...)` does an in-place modification of the airfoil). This will replace the default 2D sectional aerodynamic data with more accurate ones specific to your airfoil.

It probably goes without saying, but: if you're optimizing while applying stability (e.g. `Cma`) constraints, be sure you set a sensible `xyz_ref` in the `AeroBuildup` constructor (this should be the CG location).

Happy optimizing!

-Peter

**Reply by chiefenne** [reply]
Hi,
I am doing a similar activitiy. Can you please explain the Cma in more detail. Does it exist as a variable?
Andy
